src/dotnet-roslyn-tools/PRTagger/PRTagger.cs

```diff
@@ -2,133 +2,256 @@
+
     /// <summary>
     /// Creates GitHub issues containing the PRs inserted into a given VS build.
     /// </summary>
-    /// <param name="vsBuild">VS build number</param>
-    /// <param name="vsCommitSha">Commit SHA for VS build</param>
-    /// <param name="settings">Authentication tokens</param>
+    /// <param name="vsBuild">VS build number. Use if the tagger is invoked to check from a specify vs build.</param>
     /// <param name="logger"></param>
-    /// <param name="cancellationToken"></param>
     /// <returns>Exit code indicating whether issue was successfully created.</returns>
     public static async Task<int> TagPRs(
-        string vsBuild,
-        string vsCommitSha,
-        RoslynToolsSettings settings,
+        RemoteConnections remoteConnections,
         ILogger logger,
-        CancellationToken cancellationToken)
+        int maxFetchingVSBuildNumber,
+        string? vsBuild)
     {
-        using var devdivConnection = new AzDOConnection(settings.DevDivAzureDevOpsBaseUri, "DevDiv", settings.DevDivAzureDevOpsToken);
-        using var dncengConnection = new AzDOConnection(settings.DncEngAzureDevOpsBaseUri, "internal", settings.DncEngAzureDevOpsToken);
-        var connections = new AzDOConnection[] { devdivConnection, dncengConnection };
-
-        // Retrieve VS repo
-        var vsRepository = await GetVSRepositoryAsync(devdivConnection.GitClient);
-
-        // Find previous VS commit SHA
-        var vsCommit = await devdivConnection.GitClient.GetCommitAsync(
-            vsCommitSha, vsRepository.Id, cancellationToken: cancellationToken).ConfigureAwait(false);
-        var previousVsCommitSha = vsCommit.Parents.First();
+        var cancellationToken = CancellationToken.None;
+        var vsBuildsAndCommitSha = ImmutableArray<(string vsBuild, string vsCommitSha, string previousVsCommitSha)>.Empty;
+        // If vsbuild is not null, only check this build.
+        if (vsBuild is not null)
+        {
+            var singleBuildAndShaToCheck =
+                await GetVsBuildAndCommitAsync(remoteConnections.DevDivConnection, logger, vsBuild, cancellationToken).ConfigureAwait(false);
+            vsBuildsAndCommitSha = ImmutableArray.Create(singleBuildAndShaToCheck);
+        }
 
+        // vsBuildsAndCommitSha is ordered from new to old.
+        // For each of the product, check if the product is changed from the newest build, keep creating issues if the product has change.
+        // Stop when
+        // 1. All the VS build has been checked.
+        // 2. If some errors happens.
+        // 3. If we found the issue with the same title has been created. It means the issue is created because the last run of the tagger.
         foreach (var product in VSBranchInfo.AllProducts)
         {
             // We currently only support creating issues for GitHub repos
             if (!product.RepoHttpBaseUrl.Contains("github.com"))
             {
-                continue;
+                return -1;
```

Comment:
This change would stop us from tagging later Products that may live in GitHub. Could log a warning and continue to the next Product.